rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,501 @@
+
+Over time, the Vector team has undertaken numerous attempts to ensure that the internal telemetry
+emitted from Vector, specifically the internal telemetry related to components, is both accurate and
+consistent. This RFC charts a path for a unit test-based mechanism that will ensure that if a
+component can be used/configured by an end user, it is unit tested thoroughly, and has passed all of
+the relevant qualifications -- Component Specification adherence, correct of actual metric values,
+etc -- that make it suitable for production use.
+
+## Context
+
+- [#8858](https://github.com/vectordotdev/vector/pull/8858) - chore: Add component specification
+- [#9308](https://github.com/vectordotdev/vector/pull/9308) - chore: Add script to surface-scan
+  components against specification
+- [#9312](https://github.com/vectordotdev/vector/pull/9312) - enhancement(unit tests): Add testing
+  for component specification features
+- [#9687](https://github.com/vectordotdev/vector/issues/9687) - Audit sources to ensure compliance
+  with the component spec
+- [#9688](https://github.com/vectordotdev/vector/issues/9688) - Audit sinks to ensure compliance
+  with the component spec
+- [#10832](https://github.com/vectordotdev/vector/issues/10832) - Update component instrumentation
+  for multiple outputs
+- [#10882](https://github.com/vectordotdev/vector/issues/10882) - Vector Internal Metrics has no
+  data for `component_errors_total`
+- [#11342](https://github.com/vectordotdev/vector/issues/11342) - Ensure all existing metric `error`
+  tags comply with updated component spec
+- [#12492](https://github.com/vectordotdev/vector/pull/12492) - enhancement(datadog service): add
+  bytes sent metrics to Datadog sinks
+- [#12572](https://github.com/vectordotdev/vector/pull/12572) - chore(observability): ensure all
+  sources are checked for component spec compliance
+- [#12668](https://github.com/vectordotdev/vector/pull/12668) - chore(observability): ensure
+  transforms are checked for component spec compliance
+- [#12755](https://github.com/vectordotdev/vector/pull/12755) - chore(observability): ensure sinks
+  are checked for component spec compliance
+- [#12768](https://github.com/vectordotdev/vector/pull/12768) - chore(observability): make
+  `RequestBuilder` provide (un)compressed payload sizes
+- [#12910](https://github.com/vectordotdev/vector/pull/12910) - chore(vector source): fix the
+  `BytesReceived` event for v2 variant
+- [#12964](https://github.com/vectordotdev/vector/issues/12964) - bug(blackhole sink): does not
+  report "network" bytes sent
+- [#14073](https://github.com/vectordotdev/vector/issues/14073) - verify `route` transform's
+  `component_discarded_events_total`
+
+## Scope
+
+### In scope
+
+- Validating components against the
+  [Component Specification](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md)
+  in terms of both events emitted as well as metrics emitted.
```

Comment:
I guess I mostly saw it as: correctness of metrics is ostensibly a desired outcome for anything we emit overall, so it's not like the Component Specification says "metrics must be correct" as if we're ever OK with emitting incorrect metrics... which is why I wrote it as a different point. And also to just drive it home rather than having a single bullet point that was a long, run-on sentence.

Reply:
:+1:  makes sense. Esp after reading the rest of the RFC.